File: 1-Basic technical skills for job interviews/3-Trees/BST_implementation.dart
class TreeNode {
  int data;
  TreeNode? left;
  TreeNode? right;

  TreeNode(int value) : data = value, left = null, right = null;
}

abstract class AbstractBST {
  void insert(int value);
  bool contains(int value);
  List<int> inorderTraversal();
  List<int> preorderTraversal();
  List<int> postorderTraversal();
}

class BSTOperations implements AbstractBST {
  TreeNode? root = null;
  List<int> order = [];

  TreeNode? _insert(TreeNode? node, int value) {
    if (node == null) {
      return TreeNode(value);
    }

    if (value < node.data) {
      node.left = _insert(node.left, value);
    } else {
      node.right = _insert(node.right, value);
    }

    return node;
  }

  bool _contains(TreeNode? node, int value) {
    if (node == null) {
      return false;
    }

    if (value == node.data) {
      return true;
    }

    if (value < node.data) {
      return _contains(node.left, value);
    } else {
      return _contains(node.right, value);
    }
  }

  void _inorderTraversal(TreeNode? node) {
    if (node == null) return;

    _inorderTraversal(node.left);
    order.add(node.data);
    _inorderTraversal(node.right);
  }

  void _preorderTraversal(TreeNode? node) {
    if (node == null) return;

    order.add(node.data);
    _preorderTraversal(node.left);
    _preorderTraversal(node.right);
  }

  void _postorderTraversal(TreeNode? node) {
    if (node == null) return;

    _postorderTraversal(node.left);
    _postorderTraversal(node.right);
    order.add(node.data);
  }

  @override
  void insert(int value) {
    root = _insert(root, value);
  }

  @override
  bool contains(int value) {
    return _contains(root, value);
  }

  @override
  List<int> inorderTraversal() {
    order.clear();
    _inorderTraversal(root);
    return order;
  }

  @override
  List<int> preorderTraversal() {
    order.clear();
    _preorderTraversal(root);
    return order;
  }

  @override
  List<int> postorderTraversal() {
    order.clear();
    _postorderTraversal(root);
    return order;
  }
}

File: 1-Basic technical skills for job interviews/3-Trees/BST_implementation.cpp
#include <bits/stdc++.h>
using namespace std;

template <typename T>
class TreeNode {
public:
    T data;
    TreeNode<T>* left;
    TreeNode<T>* right;

    TreeNode(T value) : data(value), left(nullptr), right(nullptr) {}
};

/// Base class for BST Operations
template<typename T>
class AbstractBST{
public:
    virtual void insert(T value) = 0;
    virtual bool contains(T value ) = 0;
    virtual vector<T> inorderTraversal() = 0;
    virtual vector<T> preorderTraversal() = 0;
    virtual vector<T> postorderTraversal() = 0;
};

template<typename T>
class BSTOperations : public AbstractBST<T>{
private:
    TreeNode<T>* root;
    vector<T>order;
    TreeNode<T>* insert(TreeNode<T>* node, T value) {
        if (node == nullptr) {
            return new TreeNode<T>(value);
        }

        if (value < node->data) {
            node->left = insert(node->left, value);
        } else if (value > node->data) {
            node->right = insert(node->right, value);
        }

        return node;
    }
    bool contains(TreeNode<T>* node, T value){
        if (node == nullptr) {
            return false;
        }

        if (value == node->data) {
            return true;
        }

        if (value < node->data) {
            return contains(node->left, value);
        } else {
            return contains(node->right, value);
        }
    }
    void  inorderTraversal (TreeNode<T>*node){
        if(node == nullptr)return;

        inorderTraversal(node->left);
        order.push_back(node->data);
        inorderTraversal(node->right);
    }
    void  preorderTraversal (TreeNode<T>*node){
        if(node == nullptr)return;

        order.push_back(node->data);
        preorderTraversal(node->left);
        preorderTraversal(node->right);
    }
    void  postorderTraversal (TreeNode<T>*node){
        if(node == nullptr)return;

        postorderTraversal(node->left);
        postorderTraversal(node->right);
        order.push_back(node->data);

    }


public:
    BSTOperations(){
        this->root = nullptr;
    }

    void insert(T value )override{
        root = insert(root,value );
    }
    bool contains(T value) override {
        return contains(root, value);
    }

    vector<T> inorderTraversal()override{
        order.clear();
        inorderTraversal(root);
        return order;
    }
    vector<T> preorderTraversal()override{
        order.clear();
        preorderTraversal(root);
        return order;
    }

    vector<T> postorderTraversal()override{
        order.clear();
        postorderTraversal(root);
        return order;
    }

};
